src/app/rpmostree-composeutil.c

```diff
@@ -469,3 +469,104 @@ rpmostree_composeutil_write_composejson (OstreeRepo  *repo,
 
   return TRUE;
 }
+
+/* Implements --write-lockfile-to.
+ * If `path` is NULL, this is a NO-OP.
+ */
+gboolean
+rpmostree_composeutil_write_lockfilejson (RpmOstreeContext  *ctx,
+                                          const char        *path,
+                                          GError           **error)
+{
+  if (!path)
+    return TRUE;
+
+  g_autoptr(GPtrArray) pkgs = rpmostree_context_get_packages (ctx);
+  g_assert (pkgs);
+
+  g_auto(GVariantBuilder) builder;
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("a{sv}"));
+
+  g_autoptr(GVariant) pkglist_v = NULL;
+  if (!rpmostree_create_pkglist_variant (pkgs, &pkglist_v, NULL, error))
+    return FALSE;
+  g_variant_builder_add (&builder, "{sv}", "packages", pkglist_v);
+
+  g_autoptr(GVariant) lock_v = g_variant_builder_end (&builder);
+  g_assert (lock_v != NULL);
+  g_autoptr(JsonNode) lock_node = json_gvariant_serialize (lock_v);
+  g_assert (lock_node != NULL);
+  glnx_unref_object JsonGenerator *generator = json_generator_new ();
+  json_generator_set_root (generator, lock_node);
+
+  char *dnbuf = strdupa (path);
+  const char *dn = dirname (dnbuf);
+  g_auto(GLnxTmpfile) tmpf = { 0, };
+  if (!glnx_open_tmpfile_linkable_at (AT_FDCWD, dn, O_WRONLY | O_CLOEXEC, &tmpf, error))
+    return FALSE;
+  g_autoptr(GOutputStream) out = g_unix_output_stream_new (tmpf.fd, FALSE);
+  /* See also similar code in status.c */
+  if (json_generator_to_stream (generator, out, NULL, error) <= 0 ||
+      (error != NULL && *error != NULL))
+    return FALSE;
+
+  /* World readable to match --write-commitid-to which uses mask */
+  if (!glnx_fchmod (tmpf.fd, 0644, error))
+    return FALSE;
+
+  if (!glnx_link_tmpfile_at (&tmpf, GLNX_LINK_TMPFILE_REPLACE, AT_FDCWD, path, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+/* compose tree accepts JSON package version lock via file;
+ * convert it to a hash table of a{sv}; suitable for further extension.
+ */
+GHashTable *
+rpmostree_composeutil_get_vlockmap (const char  *path,
+                                    GError     **error)
+{
+  g_autoptr(JsonParser) parser = json_parser_new_immutable ();
+  if (!json_parser_load_from_file (parser, path, error))
+    {
+      g_prefix_error (error, "Loading %s: ", path);
+      return NULL;
+    }
+
+  JsonNode *metarootval = json_parser_get_root (parser);
+  g_autoptr(GVariant) jsonmetav = json_gvariant_deserialize (metarootval, "a{sv}", error);
+  if (!jsonmetav)
+    {
+      g_prefix_error (error, "Parsing %s: ", path);
+      return NULL;
+    }
+
+  g_autoptr(GVariant) value = g_variant_lookup_value (jsonmetav, "packages", G_VARIANT_TYPE ("av"));
```

Comment:
I tried, it doesn't work. The parsed json is like this
```
Read: {'packages': <[<[<'acl-2.2.53-2.fc29.x86_64'>, <'sha256:0a7446de7de962fbbfdb204da0a7aef70cc8bd4ee4b2c03b0d730bf62b9a2223'>]>, <[<'atomic-registries-1.22.1-27.gitb507039.fc29.x86_64'>, <'sha256:19fdd5bcb79bc84c4c3d17d63d2ee39173d3d2504ea11ed9f0b7d0d0cf86550c'>]>, <[<'audit-libs-3.0-0.7.20190326git03e7489.fc29.x86_64'>, <'sha256:a907544ffac01209ce49026a0ebedef58d62639aa9d234537577c72d458cc850'>]>, <[<'basesystem-11-6.fc29.noarch'>, <'sha256:e4b86fe4f119873e8705dae47163f4afb6ccf8885426e04d960ff18eed7320bb'>]>, <[<'bash-4.4.23-6.fc29.x86_64'>, <'sha256:1899be2cb4a4793d6a6c62285a6b77f5c22eb5534e76d94d2a5d421cf2aa59ad'>]>,
```
as you can see, everything is wrapped inside a GVariant. Unless I'm missing something.

Reply:
Ahh hmm, yeah I see what you mean. Seems like that's just how `json_gvariant_deserialize` works. Which makes sense, since it can't know that the JSON is following some particular scheme, so it just wraps all leaf nodes in the JSON into a GVariant. Yuck. So, I do think that parsing with serde would make this much nicer and give us scheme checking as well, but happy to stick with this for now!